### PR TITLE
 feat: Update Customer Details API

### PR DESCRIPTION
**Pull request by SCCSMARTCODE**
## Description

This pull request introduces an API endpoint to update customer details, allowing administrators to modify customer information efficiently and securely. This update ensures that customer data remains accurate and up-to-date within the system.

## Related Issue (Link to issue ticket)
  [Issue #89: Update Customer Details API](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/89)

## Motivation and Context

This feature is crucial for maintaining accurate customer information, which is vital for effective customer management and communication. By allowing admins to update customer details, the system enhances its data integrity and overall user experience.

## How Has This Been Tested?

- **Unit Testing**: Comprehensive unit tests have been written to verify that:
  - Customer details are correctly updated when valid data is provided.
  - The system returns appropriate error responses for invalid customer IDs.
  - Unauthorized users are prevented from updating customer details.
  
- **Manual Testing**: The endpoint was manually tested using Postman to ensure proper functionality and response handling.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

---

### Acceptance Criteria

1. **Endpoint Definition:**
   - API endpoint to update customer details is implemented.

2. **Functionality:**
   - Admins can update customer details.
   - Only authorized users (admins) can perform updates.

3. **Implementation Steps:**
   - **Define API Endpoint**: The endpoint for updating customer details has been defined.
   - **Implement Logic**: Logic for updating customer information in the database has been implemented.
   - **Handle Responses**: Both success and error responses, including for invalid customer ID and unauthorized access, have been handled.
   - **Write Tests**: Tests have been written and validated to ensure functionality.

### Purpose

The endpoint aims to provide a secure and efficient way for administrators to update customer details, thereby improving the system's accuracy and data management capabilities.

### Requirements

- [x] The API follows RESTful standards.
- [x] Authorization checks are in place to restrict access to admins.
- [x] Validation for incoming data is implemented.
- [x] Appropriate success and error responses are handled.

### Expected Outcome

- Successful implementation of the endpoint.
- Secure and efficient update of customer details by admins.
- Appropriate success and error responses based on the request.
- Well-tested endpoint ensuring reliability and correctness.

### Example API Endpoint

- **Endpoint:** `PUT /api/v1/customers/:customerId`

**Request Example:**

```json
{
  "firstName": "John",
  "lastName": "Doe",
  "email": "john.doe@example.com",
  "phone": "+1234567890",
  "address": "123 Main St, Anytown, USA"
}
```

**Successful Response Example:**

```json
{
  "status": "success",
  "status_code": 200,
  "message": "Customer details updated successfully.",
  "customer": {
    "id": "customerId",
    "firstName": "John",
    "lastName": "Doe",
    "email": "john.doe@example.com",
    "phone": "+1234567890",
    "address": "123 Main St, Anytown, USA"
  }
}
```

**Error Response Examples:**

1. **Invalid Customer ID:**
   ```json
   {
     "status": "error",
     "message": "Invalid customer ID.",
     "status_code": 400
   }
   ```

2. **Unauthorized Access:**
   ```json
   {
     "status": "error",
     "message": "Unauthorized access.",
     "status_code": 401
   }
   ```

### Unit Testing

- [x] Tests for successful updates with valid data and customer ID.
- [x] Tests for error handling with invalid customer ID.
- [x] Tests for authorization checks, ensuring only admins can update customer details.

### CODE REVIEW

File: api/v1/routes/customers.py
Comment (joboy-dev) [inline api/v1/routes/customers.py:1]: Making db queries in routes are not allowed. Check services folder on how to use services

File: api/v1/services/user.py
Comment (joboy-dev) [inline api/v1/services/user.py:1]: Why are you creating adkin access token??
There's already a create access token function that works for everything

**Review by joboy-dev — CHANGES_REQUESTED** [review]
Make changes

**Comment by CodewithSegNet** [comment]
also follow proper PR naming conventions. eg, feat: Update Customer …..

**Review by CodewithSegNet — COMMENTED** [review]
also follow proper PR naming conventions. eg, feat: Update Customer …..

**Review by CodewithSegNet — CHANGES_REQUESTED** [review]
also follow proper PR naming conventions. eg, feat: Update Customer …..

File: api/v1/routes/customers.py
Comment (SCCSMARTCODE) [reply api/v1/routes/customers.py:1]: This is already fixed

File: api/v1/services/user.py
Comment (SCCSMARTCODE) [reply api/v1/services/user.py:1]: I did this when the `get_current_user`  function only works with username but now that is works well with user_id and username 

i already remove the function

File: api/utils/dependencies.py
Comment (joboy-dev) [inline api/utils/dependencies.py:1]: There is already a get_current_super_admin in user service

File: api/v1/services/customers.py
Comment (joboy-dev) [inline api/v1/services/customers.py:1]: Rewrite this service file to the right format
Use a class

**Review by joboy-dev — CHANGES_REQUESTED** [review]
Remove .idea and resolve merge conflicts

**Comment by SCCSMARTCODE** [comment]
done

File: api/utils/dependencies.py
Comment (SCCSMARTCODE) [reply api/utils/dependencies.py:1]: dpne

File: api/v1/services/customers.py
Comment (SCCSMARTCODE) [reply api/v1/services/customers.py:1]: done

**Pull request closed without merging**